flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/catalog/DefaultSchemaResolver.java

```diff
@@ -0,0 +1,328 @@
+
+        // validate time attribute
+        final String timeColumn = watermarkSpec.getColumnName();
+        validateTimeColumn(timeColumn, inputColumns);
+
+        // resolve watermark expression
+        final ResolvedExpression watermarkExpression;
+        try {
+            watermarkExpression =
+                    resolveExpression(inputColumns, watermarkSpec.getWatermarkExpression());
+        } catch (Exception e) {
+            throw new ValidationException(
+                    String.format(
+                            "Invalid expression for watermark '%s'.", watermarkSpec.toString()),
+                    e);
+        }
+        validateWatermarkExpression(watermarkExpression.getOutputDataType().getLogicalType());
+
+        return Collections.singletonList(
+                new WatermarkSpec(watermarkSpec.getColumnName(), watermarkExpression));
+    }
+
+    private void validateTimeColumn(String columnName, List<Column> columns) {
+        final Optional<Column> timeColumn =
+                columns.stream().filter(c -> c.getName().equals(columnName)).findFirst();
+        if (!timeColumn.isPresent()) {
+            throw new ValidationException(
+                    String.format(
+                            "Invalid column name '%s' for rowtime attribute in watermark declaration. Available columns are: %s",
+                            columnName,
+                            columns.stream().map(Column::getName).collect(Collectors.toList())));
+        }
+        final LogicalType timeFieldType = timeColumn.get().getDataType().getLogicalType();
+        if (!hasRoot(timeFieldType, LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)
+                || getPrecision(timeFieldType) != 3) {
+            throw new ValidationException(
+                    "Invalid data type of time field for watermark definition. "
+                            + "The field must be of type TIMESTAMP(3) WITHOUT TIME ZONE.");
+        }
+        if (isProctimeAttribute(timeFieldType)) {
+            throw new ValidationException(
+                    "A watermark can not be defined for a processing-time attribute.");
+        }
+    }
+
+    private void validateWatermarkExpression(LogicalType watermarkType) {
+        if (!hasRoot(watermarkType, LogicalTypeRoot.TIMESTAMP_WITHOUT_TIME_ZONE)
+                || getPrecision(watermarkType) != 3) {
+            throw new ValidationException(
+                    "Invalid data type of expression for watermark definition. "
+                            + "The field must be of type TIMESTAMP(3) WITHOUT TIME ZONE.");
+        }
+    }
+
+    /** Updates the data type of columns that are referenced by {@link WatermarkSpec}. */
+    private List<Column> adjustRowtimeAttributes(
+            List<WatermarkSpec> watermarkSpecs, List<Column> columns) {
+        return columns.stream()
+                .map(
+                        column -> {
```

Comment:
nit: with the rather strict max columns, maybe extracting the function we map over into a separate method could improve readability a bit.